### PR TITLE
Mlc 292 model error analysis

### PR DESCRIPTION
**Pull request by noamzbr**
Add check to segment model performance per feature

### CODE REVIEW

File: deepchecks/checks/performance/model_error_analysis.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+              np.log(y_pred + np.finfo(float).eps)).sum(axis=1)
+
+
+def per_sample_mse(y_true, y_pred):
+    return (y_true - y_pred) ** 2
+
+
+def create_error_model(dataset: Dataset, random_state=42):
+    cat_features = dataset.cat_features
+    numeric_features = [num_feature for num_feature in dataset.features if num_feature not in cat_features]
```
Comment (matanper) [inline deepchecks/checks/performance/model_error_analysis.py:300]: maybe using set is more elegant `set(dataset.features) - set(cat_features)`

File: deepchecks/checks/performance/model_error_analysis.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+        error_fi.sort_values(ascending=False, inplace=True)
+
+        n_samples_display = min(self.n_display_samples, len(test_dataset))
+        display_data = test_dataset.data
+        error_col_name = 'Deepchecks model error'
+        display_error = pd.Series(error_model_predicted, name=error_col_name, index=test_dataset.data.index)
+
+        display = []
+        value = {}
+        self._scorer_name = scorer.name
```
Comment (matanper) [inline deepchecks/checks/performance/model_error_analysis.py:154]: this should be avoided. we don't want to cache things on the check. put it in the check result

File: deepchecks/checks/performance/model_error_analysis.py [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+        weak_color = '#d74949'
```
Comment (matanper) [inline deepchecks/checks/performance/model_error_analysis.py:155]: why this is not in a dict?

File: deepchecks/checks/performance/model_error_analysis.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+                cum_sum_ratio = error_per_segment_ser['count'].cumsum() / error_per_segment_ser['count'].sum()
+
+                # Partition data into two groups - weak and ok:
+
+                in_segment_indicis = cum_sum_ratio <= self.min_segment_size
+                weak_categories = error_per_segment_ser.index[in_segment_indicis]
+                ok_categories = error_per_segment_ser.index[~in_segment_indicis]
+
+                # Calculate score for each group and assign label and color
+                ok_name_feature, segment1_details = get_segment_details(model, scorer, test_dataset, data,
```
Comment (matanper) [inline deepchecks/checks/performance/model_error_analysis.py:178]: maybe better name for the segments? like `ok_segment` and `weak_segment`?

File: deepchecks/checks/performance/model_error_analysis.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+        error_fi = calculate_feature_importance(error_model, test_dataset,
+                                                permutation_kwargs={'random_state': self.random_state})
+        error_fi.sort_values(ascending=False, inplace=True)
+
+        n_samples_display = min(self.n_display_samples, len(test_dataset))
+        display_data = test_dataset.data
+        error_col_name = 'Deepchecks model error'
+        display_error = pd.Series(error_model_predicted, name=error_col_name, index=test_dataset.data.index)
+
+        display = []
```
Comment (matanper) [inline deepchecks/checks/performance/model_error_analysis.py:152]: might be better to use plotly subplots:
https://plotly.com/python/subplots/

File: deepchecks/checks/performance/model_error_analysis.py [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+                if len(result[feature]) < 2:
+                    continue
+                performance_diff = \
+                    abs(result[feature]['segment1']['score'] - result[feature]['segment2']['score']) / \
+                    abs(max(result[feature]['segment1']['score'], result[feature]['segment2']['score']))
+                if performance_diff > max_ratio_change:
+                    fails.append(feature)
+
+            if fails:
+                msg = f'Segmentation of error by the features: {", ".join(sorted(fails))} resulted in percent change' \
```
Comment (matanper) [inline deepchecks/checks/performance/model_error_analysis.py:262]: I think this message can be shorthened. remember it's not a standalone message but it's side by side to the condition name, so it need to complement it

File: deepchecks/checks/performance/model_error_analysis.py (hunk 1 above)
Comment (noamzbr) [reply deepchecks/checks/performance/model_error_analysis.py:300]: want the order to be consistent

File: deepchecks/checks/performance/model_error_analysis.py (hunk 4 above)
Comment (noamzbr) [reply deepchecks/checks/performance/model_error_analysis.py:178]: at this stage I don't know which one is ok and which one is weak

File: deepchecks/checks/performance/model_error_analysis.py [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+
+    def __init__(
+            self,
+            max_features: int = 3,
+            min_feature_contribution: float = 0.15,
+            min_error_model_score: float = 0.5,
+            min_segment_size: float = 0.05,
+            alternative_scorer: Dict[str, Callable] = None,
+            n_samples: int = 50_000,
+            n_display_samples: int = 5_000,
```
Comment (nirhutnik) [inline deepchecks/checks/performance/model_error_analysis.py:68]: well these aren't numbers

File: deepchecks/checks/performance/model_error_analysis.py [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+        test_dataset = test_dataset.sample(self.n_samples, random_state=self.random_state)
+
+        # Create scoring function, used to calculate the per sample model error
+        if task_type == ModelType.REGRESSION:
+            def scoring_func(dataset):
+                return per_sample_mse(dataset.label_col, model.predict(dataset.features_columns))
+        else:
+            def scoring_func(dataset):
+                return per_sample_binary_cross_entropy(dataset.label_col,
+                                                       model.predict_proba(dataset.features_columns))
```
Comment (nirhutnik) [inline deepchecks/checks/performance/model_error_analysis.py:125]: Why complicate this so much? Why create a scoring_func function in an if-else condition, if its only a simple wrapper, and its only used in the next lines of code?

File: deepchecks/checks/performance/model_error_analysis.py [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+
+        train_scores = scoring_func(train_dataset)
+        test_scores = scoring_func(test_dataset)
+
+        # Create and fit model to predict the per sample error
+        error_model = create_error_model(train_dataset, random_state=self.random_state)
```
Comment (nirhutnik) [inline deepchecks/checks/performance/model_error_analysis.py:131]: maybe consider a different name? maybe error_regressor? I think that "error_model" is too similar to "model_error" and that's a bit confusing. In the "create_..." func as well

File: deepchecks/checks/performance/model_error_analysis.py (hunk 2 above)
Comment (nirhutnik) [inline deepchecks/checks/performance/model_error_analysis.py:154]: Why put as part of self if only used in this function? Also, weird for it to be outside of init.

File: deepchecks/checks/performance/model_error_analysis.py [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+        n_samples_display = min(self.n_display_samples, len(test_dataset))
+        display_data = test_dataset.data
+        error_col_name = 'Deepchecks model error'
+        display_error = pd.Series(error_model_predicted, name=error_col_name, index=test_dataset.data.index)
+
+        display = []
+        value = {}
+        self._scorer_name = scorer.name
+        weak_color = '#d74949'
+        ok_color = colors['Test']
```
Comment (nirhutnik) [inline deepchecks/checks/performance/model_error_analysis.py:156]: Why not add the weak color to the colors dict? or anyway, put this constant outside?

File: deepchecks/checks/performance/model_error_analysis.py [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+
+        headnote = """<span>
+            The following graphs show the distribution of error for top features that are most useful for distinguishing
+            high error samples from low error samples.
+        </span>"""
+        display = [headnote] + display if display else None
+
+        return CheckResult(value, display=display)
+
+    def add_condition_segments_ratio_performance_change_not_greater_than(self, max_ratio_change: float = 0.05):
```
Comment (nirhutnik) [inline deepchecks/checks/performance/model_error_analysis.py:242]: maybe a more intuitive name?
"relative difference"?

File: deepchecks/checks/performance/model_error_analysis.py [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+                        color_map = {segment1_text: weak_color, segment2_text: ok_color}
+                else:
+                    color_col = data[error_col_name]
+                    color_map = None
+                display.append(px.scatter(data, x=feature, y=error_col_name, color=color_col,
+                                          title=f'Segmentation of error by {feature}',
+                                          labels={error_col_name: 'model error'},
+                                          color_discrete_map=color_map))
+
+            value[feature] = {'segment1': segment1_details or None, 'segment2': segment2_details or None}
```
Comment (nirhutnik) [inline deepchecks/checks/performance/model_error_analysis.py:230]: Why not name segments as "high error" and "low error" segments? Will also help with max() in the condition

File: deepchecks/checks/performance/model_error_analysis.py (hunk 11 above)
Comment (nirhutnik) [reply deepchecks/checks/performance/model_error_analysis.py:242]: or "relative decrease"?

File: deepchecks/checks/performance/model_error_analysis.py (hunk 6 above)
Comment (nirhutnik) [inline deepchecks/checks/performance/model_error_analysis.py:262]: percent change is a weird phrase IMO. "change percentage"?
Would change the phrasing here to maybe "Segmentation of error by the features [] resulted in change in {} larger than {}%". Although that's also not that great. Maybe we can try f2f if you agree with me

File: deepchecks/checks/performance/model_error_analysis.py [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+
+            if fails:
+                msg = f'Segmentation of error by the features: {", ".join(sorted(fails))} resulted in percent change' \
+                      f' in {self._scorer_name} larger than {format_percent(max_ratio_change)}.'
+                return ConditionResult(False, msg, category=ConditionCategory.WARN)
+            else:
+                return ConditionResult(True, category=ConditionCategory.WARN)
+
+        return self.add_condition(f'The percent change between the performance of detected segments must'
+                                  f' not exceed {format_percent(max_ratio_change)}', condition)
```
Comment (nirhutnik) [inline deepchecks/checks/performance/model_error_analysis.py:269]: Again, I think the phrasing is a bit complicated

File: deepchecks/checks/performance/model_error_analysis.py [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+        transformers=[
+            ('num', numeric_transformer, numeric_features),
+            ('cat', categorical_transformer, cat_features),
+        ]
+    )
+
+    return Pipeline(steps=[
+        ('preprocessing', preprocessor),
+        ('model', RandomForestRegressor(max_depth=4, n_jobs=-1, random_state=random_state))
+    ])
```
Comment (nirhutnik) [inline deepchecks/checks/performance/model_error_analysis.py:317]: Maybe make sure that the column issue from unused_features does not happen here as well?

**Review by nirhutnik — CHANGES_REQUESTED** [review]

File: deepchecks/checks/performance/model_error_analysis.py (hunk 7 above)
Comment (noamzbr) [reply deepchecks/checks/performance/model_error_analysis.py:68]: It's legit in python, as the great @matanper taught me

File: deepchecks/checks/performance/model_error_analysis.py (hunk 8 above)
Comment (noamzbr) [reply deepchecks/checks/performance/model_error_analysis.py:125]: It's simple but it's better than just doing it all in the if else IMO

File: deepchecks/checks/performance/model_error_analysis.py (hunk 12 above)
Comment (noamzbr) [reply deepchecks/checks/performance/model_error_analysis.py:230]: At this stage, I don't know yet if it's high or low. Need to check in that if.

File: deepchecks/checks/performance/model_error_analysis.py (hunk 14 above)
Comment (noamzbr) [reply deepchecks/checks/performance/model_error_analysis.py:317]: GOOD CATCH!

File: deepchecks/checks/performance/model_error_analysis.py (hunk 5 above)
Comment (noamzbr) [reply deepchecks/checks/performance/model_error_analysis.py:152]: Later on perhaps(:

File: deepchecks/checks/performance/model_error_analysis.py (hunk 3 above)
Comment (noamzbr) [reply deepchecks/checks/performance/model_error_analysis.py:155]: Not really more convenient

**Comment by noamzbr** [comment]
<p>resolved</p>

---

 <a href='https://app.reviewnb.com/deepchecks/deepchecks/pull/378/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by noamzbr** [comment]
<p>seems more "flow like" here</p>

---

 <a href='https://app.reviewnb.com/deepchecks/deepchecks/pull/378/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by noamzbr** [comment]
<p>splitted</p>

---

 <a href='https://app.reviewnb.com/deepchecks/deepchecks/pull/378/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: deepchecks/checks/performance/model_error_analysis.py (hunk 7 above)
Comment (nirhutnik) [reply deepchecks/checks/performance/model_error_analysis.py:93]: Cool.

File: deepchecks/checks/performance/model_error_analysis.py (hunk 8 above)
Comment (nirhutnik) [reply deepchecks/checks/performance/model_error_analysis.py:151]: Disagree, no need to create an instance of a wrapper if its only used once and directly afterwards.
But not going to fight on it. Will change it behind your back sometime

File: deepchecks/checks/performance/model_error_analysis.py (hunk 12 above)
Comment (nirhutnik) [reply deepchecks/checks/performance/model_error_analysis.py]: Yes, but why not insert it into "low" and "high" to begin with when you're doing the tree_partitioner thing? why keep it as segment1 and 2? for what benefit?

**Review by nirhutnik — APPROVED** [review]

File: deepchecks/checks/performance/model_error_analysis.py (hunk 8 above)
Comment (noamzbr) [reply deepchecks/checks/performance/model_error_analysis.py:151]: Used twice!
```
train_scores = scoring_func(train_dataset)
test_scores = scoring_func(test_dataset)
```

File: deepchecks/checks/performance/model_error_analysis.py (hunk 12 above)
Comment (noamzbr) [reply deepchecks/checks/performance/model_error_analysis.py]: oh now I get it, yeah could have changed it here...

**Pull request merged**